src/discord_objects.py

```diff
@@ -0,0 +1,115 @@
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Self
+
+
+@dataclass
+class DiscordUser:
+    id: int
+    avatar: str
+    user_name: str
+    global_name: str
+
+    @classmethod
+    def from_dict(cls, user: dict) -> Self:
+        return cls(
+            id = user["id"],
+            avatar = user["avatar"],
+            user_name = user["username"],
+            global_name = user["global_name"],
+        )
+
+    def get_user_name(self) -> str:
+        return self.global_name if self.global_name else self.user_name
+
+    def get_avatar_url(self, size: int = 128) -> str:
+        avatar_url = ""
+        if self.avatar:
+            avatar_url = f"https://cdn.discordapp.com/avatars/{self.id}/{self.avatar}.webp?size={size}"
+        return avatar_url
+
+@dataclass
+class DiscordFriend:
+    user: DiscordUser
+    nickname: str
+
+    @classmethod
+    def from_dict(cls, friend: dict) -> Self:
+        return cls(
+            user = DiscordUser.from_dict(friend["user"]),
+            nickname = friend["nickname"],
+        )
+
+    def get_friend_name(self) -> str:
+        return self.nickname if self.nickname else self.user.get_user_name()
+
+@dataclass
+class DiscordMessage:
+    id: int
+    author: DiscordUser
+    content: str
+    timestamp: datetime
+
+    @classmethod
+    def from_dict(cls, message: dict) -> Self:
+        if not (content := message["content"]):
+            content = "[(call/image/other)]"
+
+        # For some reason, messages can use two, slightly different, timestamp formats.
+        time_str = message["timestamp"]
+        if len(time_str) == 32:
+            timestamp = datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S.%f%z")
+        else:
+            timestamp = datetime.strptime(time_str, "%Y-%m-%dT%H:%M:%S%z")
+
+        return cls(
+            id=message["id"],
+            author=DiscordUser.from_dict(message["author"]),
+            content=content,
+            timestamp=timestamp.astimezone(),
+        )
+
+@dataclass
+class DiscordChannel:
+    id: int
+    type: int
+    name: str | None
```

Comment:
Does this mean that the discord channel doesn't have to have a name? 
Does `str | None` mean that it can be of type string or None?

Reply:
Per the Discord's developer docs, a channel object does not always have a `name` field: https://discord.com/developers/docs/resources/channel#channel-object-channel-structure
So yes, is `str | None` so when trying to get the name field from a channel object `dict.get("name")` can return the string of the name, or None if the channel object doesn't have that field.
https://docs.python.org/3.11/library/stdtypes.html#types-union